pkg/types/cluster.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+import (
+	"fmt"
+	"net/url"
+	"sort"
+	"strings"
+)
+
+type Cluster map[string]URLs
+
+// NewCluster returns a Cluster instantiated from the given cluster token
+// and cluster string, by parsing members from a set of discovery-formatted
+// names-to-IPs, like:
+// mach0=http://1.1.1.1,mach0=http://2.2.2.2,mach1=http://3.3.3.3,mach2=http://4.4.4.4
+func NewCluster(s string) (Cluster, error) {
+	cl := make(Cluster)
+	v, err := url.ParseQuery(strings.Replace(s, ",", "&", -1))
+	if err != nil {
+		return nil, err
+	}
+	for name, urls := range v {
+		if len(urls) == 0 || urls[0] == "" {
+			return nil, fmt.Errorf("Empty URL given for %q", name)
+		}
+		us, err := NewURLs(urls)
+		if err != nil {
+			return nil, err
+		}
+		cl[name] = us
+	}
+	return cl, nil
+}
+
+func (c Cluster) String() string {
+	kvs := make([]string, 0)
+	for name, peer := range c {
+		for _, url := range peer {
+			kvs = append(kvs, fmt.Sprintf("%s=%s", name, url.String()))
+		}
+	}
+	sort.Strings(kvs)
+	return strings.Join(kvs, ",")
+}
+
+func (c Cluster) Set(name string, urls URLs) { c[name] = urls }
+
+func (c Cluster) Get(name string) URLs { return c[name] }
+
+// URLs returns a list of all peer addresses.
+// The returned list is sorted in ascending lexicographical order.
+func (c Cluster) URLs() []string {
+	urls := make([]string, 0)
+	for _, peer := range c {
+		for _, url := range peer {
+			urls = append(urls, url.String())
+		}
+	}
+	sort.Strings(urls)
+	return urls
+}
+
+// Validate ensures that there is no identical urls in the cluster peer list
+func (c Cluster) Validate() error {
+	urlMap := make(map[string]bool)
+	for _, peer := range c {
+		for _, url := range peer {
+			u := url.String()
+			if urlMap[u] {
+				return fmt.Errorf("duplicate url %v in cluster config", url)
```

Comment:
what is a cluster config? URL or url? (see line 38)

`duplicate URL %q found in cluster`